datafusion/physical-plan/src/sorts/builder.rs

```diff
@@ -42,15 +37,15 @@ pub struct BatchBuilder {
     reservation: MemoryReservation,
 
     /// The current [`BatchCursor`] for each stream
-    cursors: Vec<BatchCursor>,
+    cursors: Vec<Option<BatchCursor<C>>>,
```

Comment:
This is indexed per stream. As such, we don't always have an ongoing cursor per each stream => hence the option.

Reply:
It seems like this structure now does some of the same things as `SortPreservingMergeStream`: https://github.com/apache/arrow-datafusion/blob/37d6bf08c948418fe6c72d072d988c2875d81e02/datafusion/physical-plan/src/sorts/merge.rs#L210-L222

Is there any way to avoid a second copy?

Reply:
All of the cursor management/ownership is moved into the SortOrderBuilder, and removed from the SortPreservingMergeStream. No second copy. Instead the division of concerns is:
* SortPreservingMergeStream == only care about merging (via loser tree)
* SortOrderBuilder == holds the cursor (being advanced), and the sort_order (being built).

This clearly demonstrates why I need to add more docs to the BatchCursor. 😅 Doing so.